spock-core/src/main/java/org/spockframework/runtime/DataIteratorFactory.java

```diff
@@ -40,6 +42,105 @@ protected Object invokeRaw(Object target, MethodInfo method, Object... arguments
         return null;
       }
     }
+
+    protected int estimateNumIterations(Object dataProvider) {
+      if (context.getErrorInfoCollector().hasErrors()) {
+        return -1;
+      }
+
+      // unbelievably, DefaultGroovyMethods provides a size() method for Iterators,
+      // although it is of course destructive (i.e. it exhausts the Iterator)
+      if (dataProvider instanceof Iterator) {
+        return -1;
+      }
+
+      Object rawSize = GroovyRuntimeUtil.invokeMethodQuietly(dataProvider, "size");
+      if (!(rawSize instanceof Number)) {
+        return -1;
+      }
+
+      int size = ((Number) rawSize).intValue();
+      if (size < 0) {
+        return -1;
+      }
+
+      return size;
+    }
+
+    protected int estimateNumIterations(Object[] dataProviders) {
+      if (context.getErrorInfoCollector().hasErrors()) {
+        return -1;
+      }
+      if (dataProviders.length == 0) {
+        return 1;
+      }
+
+      int result = Integer.MAX_VALUE;
+      for (Object prov : dataProviders) {
+        int size = estimateNumIterations(prov);
+        if (size < 0 || size >= result) {
+          continue;
+        }
+        result = size;
+      }
+
+      return result == Integer.MAX_VALUE ? -1 : result;
+    }
+
+    protected boolean haveNext(Iterator<?>[] iterators, List<DataProviderInfo> dataProviderInfos) {
+      boolean result = true;
+
+      for (int i = 0; i < iterators.length; i++)
+        try {
+          boolean hasNext = iterators[i].hasNext();
+          if (i == 0) {
+            result = hasNext;
+          } else if (result != hasNext) {
+            DataProviderInfo provider = dataProviderInfos.get(i);
+            supervisor.error(context.getErrorInfoCollector(), new ErrorInfo(provider.getDataProviderMethod(),
+              createDifferentNumberOfDataValuesException(provider, hasNext)));
+            return false;
+          }
+
+        } catch (Throwable t) {
+          supervisor.error(context.getErrorInfoCollector(), new ErrorInfo(dataProviderInfos.get(i).getDataProviderMethod(), t));
+          return false;
+        }
+
+      return result;
+    }
+
+    protected Iterator<?> createIterator(Object dataProvider, DataProviderInfo dataProviderInfo) {
```

Comment:
I don't think adding `@Nullable` is too appropriate here.
Yes, the method might return `null`.
But within all these compiler things, we have the error collector, collecting and properly propagating errors.
This method only returns `null` if there was already an error collected before or if there was an error collected during execution.
And all subsequent methods are expected to not do anything if there was already an error collected, which is also the reason this method returns `null` after all.
If it would return `null` without collecting an error, I would agree.